src/allmydata/web/storage.py

```diff
@@ -260,8 +284,52 @@ def add(*pieces):
 
         if last["corrupt-shares"]:
             add("Corrupt shares:",
-                T.ul[ [T.li[ ["SI %s shnum %d" % corrupt_share
+                T.ul( (T.li( ["SI %s shnum %d" % corrupt_share
                               for corrupt_share in last["corrupt-shares"] ]
-                             ]]])
+                             ))))
+
+        return tag(p)
+
+    def format_recovered(self, sr, a):
+        def maybe(d):
+            if d is None:
+                return "?"
+            return "%d" % d
+        return "%s shares, %s buckets (%s mutable / %s immutable), %s (%s / %s)" % \
+               (maybe(sr["%s-shares" % a]),
+                maybe(sr["%s-buckets" % a]),
+                maybe(sr["%s-buckets-mutable" % a]),
+                maybe(sr["%s-buckets-immutable" % a]),
+                abbreviate_space(sr["%s-diskbytes" % a]),
+                abbreviate_space(sr["%s-diskbytes-mutable" % a]),
+                abbreviate_space(sr["%s-diskbytes-immutable" % a]),
+                )
+
+class StorageStatus(MultiFormatResource):
+    def __init__(self, storage, nickname=""):
+        super(StorageStatus, self).__init__()
+        self.storage = storage
+        self.nickname = nickname
+
+    def render_HTML(self, req):
+        return renderElement(req, StorageStatusElement(self.storage, self.nickname))
+
+    def render_JSON(self, req):
+        req.setHeader("content-type", "text/plain")
+        d = {"stats": self.storage.get_stats(),
+             "bucket-counter": self.storage.bucket_counter.get_state(),
+             "lease-checker": self.storage.lease_checker.get_state(),
+             "lease-checker-progress": self.storage.lease_checker.get_progress(),
+             }
+        return json.dumps(d, indent=1) + "\n"
+
+    def renderSynchronously(self):
+        # to appease the test suite.
+        elem = StorageStatusElement(self.storage, self.nickname)
+        result = []
+        flattenString(None, elem).addCallback(result.append)
+        return result[0]
```

Comment:
This code should just be in the test suite.  It's okay for a test to use `flattenString` but it's less legitimate anywhere else.  Also, in the test suite, you have the handy `TestCase.successResultOf` helper that lets you skip the `[].append` trick.

Reply:
This is done in d23bc70 and a9db5b6.